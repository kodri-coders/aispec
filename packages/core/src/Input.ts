import { type XMLBaseConstructor } from './XMLBase';

interface IInputConstructor extends XMLBaseConstructor {
  name: string;
  description: string;
  schema: string;
}

export class Input {
  name: string;
  description: string;
  schema: string;

  constructor(public input: IInputConstructor) {
    this.name = input.name;
    this.description = input.description;
    this.schema = JSON.parse(input.schema);
  }
};
