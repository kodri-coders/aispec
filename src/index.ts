import { nullLiteral } from '@babel/types';
import { XMLParser, XMLBuilder } from 'fast-xml-parser';
import fs from 'fs';
import { EventEmitter } from 'stream';
import { generateText, tool } from 'ai';
import { z } from 'zod';

interface XMLBase {
  "@ref"?: string;
}

interface IAssistantConstructor extends XMLBase {
  '@id'?: string;
  name?: string;
  description?: string;
  skills?: {
    skill: any[];
  };
  workflows?: {
    workflow: any[];
  };
}

interface ISkillConstructor extends XMLBase {
  '@id': string;
  name: string;
  description: string;
  workflows?: {
    workflow: any[];
  };
}

interface IWorkflowConstructor extends XMLBase {
  '@id': string;
  name: string;
  description: string;
  steps: {
    step: {
      name: string;
      description: string;
      input: any;
      prompt: string;
      outputs: any;
    }[];
  };
}

interface IStepConstructor extends XMLBase {
  name: string;
  description: string;
  input: any;
  prompt: string;
  outputs: any;
  $ref?: string;
}

interface IInputConstructor extends XMLBase {
  name: string;
  description: string;
  schema: string;
}

interface IOutputConstructor extends XMLBase {
  name: string;
  description: string;
  schema: string;
}

const parser = new XMLParser({
  ignoreAttributes: false,
  attributeNamePrefix: '@',
  isArray: (name) => ['skill', 'workflow', 'step'].includes(name)
});
interface BaseNode {
  '@ref'?: string;
  [key: string]: any
}
class XMLBase {
  node?: any
  [key: string]: any; // Allows any string key to access properties of type any

  constructor(
    node: BaseNode,
  ) {

    if(node['@ref']){
      this.node = XMLBase.loadXMLFile(node['@ref']);
      if(Array.isArray(this.node)){
        this.node = this.node[0];
      }
      return
    }
    this.node = node;
  }
  toString(): string {
    // reutrn xml
    const builder = new XMLBuilder({
      ignoreAttributes: false,
      attributeNamePrefix: '@',
      format: true,
    });
    const node: any = {};
    Object.keys(this.node).forEach(key => {
      if (this[key]) {
        node[key] = this.node[key];
      }
    });
    return builder.build(node);
  }
  static loadXML(xml: string): any {
    const parsed = parser.parse(xml);
    // get current constructor class and load xml into it
    const topElement = Object.keys(parsed)[0];
    // instantiate based on the class which extends XMLBase
    //@ts-ignore
    return parsed[topElement]
  }
  static loadXMLFile(xmlPath: string): any {
    const xml = fs.readFileSync(xmlPath, 'utf-8');
    return XMLBase.loadXML(xml);
  }
}

export class Assistant extends XMLBase {
  name?: string;
  description?: string;
  workflows?: Workflow[] = [];
  skills?: Skill[] = [];

  constructor(assistant: IAssistantConstructor | XMLBase) {
    super(assistant);
    this.name = this.node.name;
    this.description = this.node.description;
    this.model = this.node.model;

    this.skills = this.node.skills?.skill.map((skill: any) => {
      return new Skill(skill, this);
    });
    
    this.workflows = this.node.workflows?.workflow.map((workflow: any) => {
      return new Workflow(workflow);
    });

  }
  findWorkflow(workflowId: string):any {
    let targetWorkflow: Workflow | undefined;
    let containingSkill: Skill | undefined;

    // Check global workflows
    targetWorkflow = this.workflows?.find(w => w.workflow['@id'] === workflowId);

    // If not found in global workflows, check skills
    if (!targetWorkflow) {
      this.skills?.forEach(skill => {
        const found = skill.workflows?.find(w => w.workflow['@id'] === workflowId);
        if (found) {
          targetWorkflow = found;
          containingSkill = skill;
        }
      });
    }

    if (!targetWorkflow) {
      throw new Error(`Workflow with id ${workflowId} not found`);
    }
    return {workflow: targetWorkflow, skill: containingSkill}
  }
  loadWorkflow(workflowId: string, input?: any): any {
    // Find the workflow and its containing skill
    const {workflow: targetWorkflow, skill: containingSkill} = this.findWorkflow(workflowId);
    const xmlObj: any = {
    assistant: {
      name: this.name,
      description: this.description,
      skills: {
        skill: []
      }
    }
  };

  // Add global skills
  if (this.skills) {
    xmlObj.assistant.skills.skill = this.skills
      .filter(s => s !== containingSkill)
      .map(s => ({ name: s.name, description: s.description }));
  }
  if(targetWorkflow) {
    xmlObj.assistant.workflows = {
      workflow: [targetWorkflow]
    }
  }
  // Add the skill containing the workflow
  if (containingSkill) {
    xmlObj.assistant.skills.skill.push(containingSkill.node);
  }

    return new WorkflowRunner(new Assistant(xmlObj), targetWorkflow)
  }
}
class WorkflowRunner extends EventEmitter{

  currentStep: number = -1;
  workflow: any;
  finished: boolean = false;
  context: any;
  history: any[] = [];
  assistant: Assistant;
  constructor(assistant: Assistant, workflow: any) {
    super();
    this.assistant = assistant;
    this.workflow = workflow;
  }
  onFinish(){
    return new Promise((resolve, reject) => {
      if(this.finished) {
        resolve(this.workflow);
        return
      }
      this.on('workflow-finished', (workflow: any) => {
        console.log('return workflow')
        resolve(workflow);

      });
    });
  }
  setContext(context: any) {
    this.context = { ...this.context, ...context};
  }
  start(input?: any): void {
    this.setContext(input);
    this.next()
  }
  callLLM(systemPrompt: string, prompt: string, tools: any, model: any): any {
    console.log('prompt', prompt)
    console.log('tools', tools)
    console.log('model', model)
  }
  generateResponse(prompt: string): any {
    const step = this.workflow.steps[this.currentStep];
    const model = step.model || this.workflow.model || this.assistant.model;
    const toolSchema = step.output.schema;
    const responseTool = {
      description: step.output.description || step.output.name,
      parameters: [
        toolSchema
      ],
      execute: async (params: any) => {
        this.setContext(params);
        this.emit('step-finished', prompt, params);
        return true;
      },
    };
    const systemPrompt = this.toXML(this.assistant.node)
    this.callLLM(systemPrompt, prompt, [responseTool], model);
  }
  toXML(node: any): string {
    const builder = new XMLBuilder({
      ignoreAttributes: false,
      attributeNamePrefix: '@',
      format: true,
    });
    return builder.build(node);
  }
  next(): void {
    this.currentStep++;
    if(this.currentStep >= this.workflow.steps.length) {
      this.emit('workflow-finished', this.workflow);
      this.finished = true;
      return 
    }
    const step = this.workflow.steps[this.currentStep];
    const prompt = step.getPrompt(this.context);
    const data = this.generateResponse(prompt);
    this.history.push({prompt, data});
    this.setContext(data);
    console.log({context: this.context})
    this.emit('step-finished', prompt, data);
  }



}
class Skill extends XMLBase {
  name: string ;
  description: string;
  workflows?: Workflow[] = [];
  constructor(public skill: ISkillConstructor, private parent?: Assistant) {
    super(skill);
    this.name = this.node.name;
    this.description = this.node.description;
    this.workflows = this.node.workflows?.workflow.map((workflow: any) => {
      return new Workflow(workflow, parent);
    });
    
    this.workflows?.forEach((workflow: any) => {
      this.parent?.workflows?.push(workflow);
    })
  }

}

class Workflow extends XMLBase {
  name: string;
  description: string;
  steps: Step[] = [];
  constructor(public workflow: IWorkflowConstructor, private parent?: Assistant) {
    super(workflow);
    this.name = this.node.name;
    this.description = this.node.description;
    this.steps = this.node.steps.step.map((step: any) => {
      return new Step(step, parent);
    });
  }
  run(input?: any): any {
    return this.steps.map((step: any) => step.runStep(input));
  }

}

class Step extends XMLBase {
  name: string;
  description: string;
  input: Input;
  prompt?: Prompt;
  output: Output;
  constructor(public step: IStepConstructor, private parent?: Assistant) {
    super(step);
    this.name = this.node.name;
    this.description = this.node.description;
    this.input = this.node.input && new Input(this.node.input);
    this.prompt = this.node.prompt && new Prompt(this.node.prompt);
    this.output = this.node.output && new Output(this.node.output);
  }
  getPrompt(input?: any): any {
    return this.prompt?.interpolate(input);
  }
}

class Prompt {
  private prompt: string;
  public variables: string[];
  constructor(prompt: string) {
    this.prompt = prompt;
    this.variables = this.extractVariables(prompt);
  }
  private extractVariables(prompt: string): string[] {
    const variables = prompt.match(/\$\{(.+?)\}/g);
    return variables || [];
  }
  interpolate(data: any): string {
    return this.prompt.replace(/\${(.+?)}/g, (match, key) => {
      let value = '()';
      key.split('.').forEach((keyPart: string) => {
        value = data[keyPart];
      });
      return value;
    });
  }
}

class Input {
  name: string;
  description: string;
  schema: string;
  constructor(public input: IInputConstructor) {
    this.name = input.name;
    this.description = input.description;
    this.schema = JSON.parse(input.schema);
  }
}

class Output {
  name: string;
  description: string;
  schema: string;
  constructor(public output: IOutputConstructor) {
    this.name = output.name;
    this.description = output.description;
    this.schema = JSON.parse(output.schema);
  }
}
