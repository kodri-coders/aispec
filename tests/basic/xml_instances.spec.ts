import { Assistant } from '../../src/index'
import * as path from 'path'
const artifactPath = path.join(__dirname, 'artifacts', 'movie-builder-assistant.xml')
const assistant = new Assistant({"@ref":artifactPath})
describe('assistant', () => {
  it('loads name and description', () => {
    expect(assistant.name).toMatchSnapshot()
    expect(assistant.description).toMatchSnapshot()
  })
  it('loads xml', () => {
    expect(assistant).toMatchSnapshot() 
  })
  it('loads workflows', () => {
    expect(assistant.workflows).toMatchSnapshot()
  })
  it('starts workflow', async () => {
    const workflowRunner = assistant.loadWorkflow('character-building')
    const mockResponses: any = {
      'Produce 2 surnames that sound good with the name: John': {
        surnames: ['Doe', 'Smith']
      },
      'Given the name: John select one of the surnames: Doe,Smith': {
        surname: 'Doe'
      },
      'Generate a character with the name: John and the surname: Doe': {
        character: 'John Doe'
      }
    }
    // mock workflow.generateResponse with jest based
    workflowRunner.generateResponse = jest.fn().mockImplementation((prompt: string) => {
      console.log('prompt', prompt)
      return mockResponses[prompt]
    })
    workflowRunner.on('step-finished', (prompt: string, data: any) => {
      console.log('step finished', prompt, data)
      workflowRunner.next()
    })
    workflowRunner.on('workflow-finished', (workflow: any) => {
      console.log('workflow finished')
    })
    const workflow = workflowRunner.start({
      name: 'John',
      surnamesLength: 2,
    })

    console.log('workflow', workflow)
    await workflowRunner.onFinish()
    expect(workflowRunner.generateResponse).toHaveBeenCalledTimes(3)
    expect(workflowRunner.history).toMatchSnapshot()
  })
})
