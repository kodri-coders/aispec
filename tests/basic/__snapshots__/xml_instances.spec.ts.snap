// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assistant loads name and description 1`] = `"Movie builder"`;

exports[`assistant loads name and description 2`] = `"Assistant 1 description"`;

exports[`assistant loads workflows 1`] = `
[
  Workflow {
    "description": "Workflow 2 description",
    "name": "Movie ideas",
    "node": {
      "@id": "movie-ideas",
      "description": "Workflow 2 description",
      "input": {
        "description": "The genre of the movie",
        "name": "Movie idea genre",
        "schema": "{
                        "type": "object",
                        "properties": {
                            "genre": {
                                "type": "string"
                            }
                        }
                    }",
      },
      "name": "Movie ideas",
      "steps": {
        "step": [
          {
            "@id": "ideas-generation",
            "output": {
              "@id": "ideas",
              "name": "Movie ideas",
              "schema": "{
                                "type": "object",
                                "properties": {
                                    "ideas": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    }
                                }
                            }",
            },
            "prompt": "generate 10 movie ideas for \${genre}. 
                    Create a story that takes place in a 
                    world with certain characters, plot lines and themes.",
          },
        ],
      },
    },
    "parent": undefined,
    "steps": [
      Step {
        "description": undefined,
        "input": undefined,
        "name": undefined,
        "node": {
          "@id": "ideas-generation",
          "output": {
            "@id": "ideas",
            "name": "Movie ideas",
            "schema": "{
                                "type": "object",
                                "properties": {
                                    "ideas": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    }
                                }
                            }",
          },
          "prompt": "generate 10 movie ideas for \${genre}. 
                    Create a story that takes place in a 
                    world with certain characters, plot lines and themes.",
        },
        "output": Output {
          "description": undefined,
          "name": "Movie ideas",
          "output": {
            "@id": "ideas",
            "name": "Movie ideas",
            "schema": "{
                                "type": "object",
                                "properties": {
                                    "ideas": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    }
                                }
                            }",
          },
          "schema": {
            "properties": {
              "ideas": {
                "items": {
                  "type": "string",
                },
                "type": "array",
              },
            },
            "type": "object",
          },
        },
        "parent": undefined,
        "prompt": Prompt {
          "prompt": "generate 10 movie ideas for \${genre}. 
                    Create a story that takes place in a 
                    world with certain characters, plot lines and themes.",
          "variables": [
            "\${genre}",
          ],
        },
        "step": {
          "@id": "ideas-generation",
          "output": {
            "@id": "ideas",
            "name": "Movie ideas",
            "schema": "{
                                "type": "object",
                                "properties": {
                                    "ideas": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    }
                                }
                            }",
          },
          "prompt": "generate 10 movie ideas for \${genre}. 
                    Create a story that takes place in a 
                    world with certain characters, plot lines and themes.",
        },
      },
    ],
    "workflow": {
      "@id": "movie-ideas",
      "description": "Workflow 2 description",
      "input": {
        "description": "The genre of the movie",
        "name": "Movie idea genre",
        "schema": "{
                        "type": "object",
                        "properties": {
                            "genre": {
                                "type": "string"
                            }
                        }
                    }",
      },
      "name": "Movie ideas",
      "steps": {
        "step": [
          {
            "@id": "ideas-generation",
            "output": {
              "@id": "ideas",
              "name": "Movie ideas",
              "schema": "{
                                "type": "object",
                                "properties": {
                                    "ideas": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    }
                                }
                            }",
            },
            "prompt": "generate 10 movie ideas for \${genre}. 
                    Create a story that takes place in a 
                    world with certain characters, plot lines and themes.",
          },
        ],
      },
    },
  },
]
`;

exports[`assistant loads xml 1`] = `
Assistant {
  "description": "Assistant 1 description",
  "model": {
    "max_tokens": 1000,
    "name": "gpt-4o",
    "temperature": 0.5,
  },
  "name": "Movie builder",
  "node": {
    "@id": "movie-builder",
    "description": "Assistant 1 description",
    "model": {
      "max_tokens": 1000,
      "name": "gpt-4o",
      "temperature": 0.5,
    },
    "name": "Movie builder",
    "skills": {
      "skill": [
        {
          "@ref": "tests/basic/artifacts/world-building-skill.xml",
        },
        {
          "description": "Building a movie scenario means
            creating a story that takes place in
            a world with certain characters, 
            plot lines and themes. 
            The world can be a fantasy world, 
            a historical period, a real world city, 
            or any other setting imaginable. 
            The story can be a romance, a thriller, 
            a sci-fi epic or any other genre. The characters can be humans, animals, aliens, or any other type of being. The plot can be a simple one, like a romance that takes place in a small town, or a complex one, like a multi-layered epic that takes place in a fictional world. The themes can be simple, like love and friendship, or complex, like the nature of reality and the human condition. The movie scenario building skill allows you to explore all of these possibilities and create a unique story that expresses your imagination and creativity.",
          "name": "Movie building skill",
        },
      ],
    },
    "workflows": {
      "workflow": [
        {
          "@id": "movie-ideas",
          "description": "Workflow 2 description",
          "input": {
            "description": "The genre of the movie",
            "name": "Movie idea genre",
            "schema": "{
                        "type": "object",
                        "properties": {
                            "genre": {
                                "type": "string"
                            }
                        }
                    }",
          },
          "name": "Movie ideas",
          "steps": {
            "step": [
              {
                "@id": "ideas-generation",
                "output": {
                  "@id": "ideas",
                  "name": "Movie ideas",
                  "schema": "{
                                "type": "object",
                                "properties": {
                                    "ideas": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    }
                                }
                            }",
                },
                "prompt": "generate 10 movie ideas for \${genre}. 
                    Create a story that takes place in a 
                    world with certain characters, plot lines and themes.",
              },
            ],
          },
        },
      ],
    },
  },
  "skills": [
    Skill {
      "description": "A skill that allows you to build a world",
      "name": "World building",
      "node": {
        "@id": "skill1",
        "description": "A skill that allows you to build a world",
        "name": "World building",
        "workflows": {
          "workflow": [
            {
              "@id": "character-building",
              "description": "A workflow that allows you to build a character with a given name",
              "input": {
                "description": "Input description",
                "name": "Human name",
                "schema": "{
            "type": "object",
            "properties": {
              "name": {
                "type": "string"
              },
              "surnamesLength": {
                "type": "number"
              }
            }
          }",
              },
              "name": "Character building",
              "steps": {
                "step": [
                  {
                    "@id": "surname-generation",
                    "description": "Surname generation step",
                    "name": "Surname generation",
                    "output": {
                      "@id": "surnames",
                      "name": "Surnames",
                      "schema": "{
                "type": "object",
                "properties": {
                  "surnames": {
                    "type": "array",
                    "items": {
                      "type": "string"
                    }
                  }
                }
              }",
                    },
                    "prompt": "Produce \${surnamesLength} surnames that sound good with the name: \${name}",
                  },
                  {
                    "@id": "surname-selection",
                    "description": "Surname selection step",
                    "name": "Surname selection",
                    "output": {
                      "@id": "surname",
                      "name": "Surname",
                      "schema": "{
                "type": "object",
                "properties": {
                  "surname": {
                    "type": "string"
                  }
                }
              }",
                    },
                    "prompt": "Given the name: \${name} select one of the surnames: \${surnames}",
                  },
                  {
                    "@id": "character-generation",
                    "description": "Character generation step",
                    "name": "Character generation",
                    "output": {
                      "@id": "character",
                      "name": "Character",
                      "schema": "{
                "type": "object",
                "properties": {
                  "character": {
                    "type": "string"
                  }
                }
              }",
                    },
                    "prompt": "Generate a character with the name: \${name} and the surname: \${surname}",
                  },
                ],
              },
            },
          ],
        },
      },
      "parent": [Circular],
      "skill": {
        "@ref": "tests/basic/artifacts/world-building-skill.xml",
      },
      "workflows": [
        Workflow {
          "description": "A workflow that allows you to build a character with a given name",
          "name": "Character building",
          "node": {
            "@id": "character-building",
            "description": "A workflow that allows you to build a character with a given name",
            "input": {
              "description": "Input description",
              "name": "Human name",
              "schema": "{
            "type": "object",
            "properties": {
              "name": {
                "type": "string"
              },
              "surnamesLength": {
                "type": "number"
              }
            }
          }",
            },
            "name": "Character building",
            "steps": {
              "step": [
                {
                  "@id": "surname-generation",
                  "description": "Surname generation step",
                  "name": "Surname generation",
                  "output": {
                    "@id": "surnames",
                    "name": "Surnames",
                    "schema": "{
                "type": "object",
                "properties": {
                  "surnames": {
                    "type": "array",
                    "items": {
                      "type": "string"
                    }
                  }
                }
              }",
                  },
                  "prompt": "Produce \${surnamesLength} surnames that sound good with the name: \${name}",
                },
                {
                  "@id": "surname-selection",
                  "description": "Surname selection step",
                  "name": "Surname selection",
                  "output": {
                    "@id": "surname",
                    "name": "Surname",
                    "schema": "{
                "type": "object",
                "properties": {
                  "surname": {
                    "type": "string"
                  }
                }
              }",
                  },
                  "prompt": "Given the name: \${name} select one of the surnames: \${surnames}",
                },
                {
                  "@id": "character-generation",
                  "description": "Character generation step",
                  "name": "Character generation",
                  "output": {
                    "@id": "character",
                    "name": "Character",
                    "schema": "{
                "type": "object",
                "properties": {
                  "character": {
                    "type": "string"
                  }
                }
              }",
                  },
                  "prompt": "Generate a character with the name: \${name} and the surname: \${surname}",
                },
              ],
            },
          },
          "parent": [Circular],
          "steps": [
            Step {
              "description": "Surname generation step",
              "input": undefined,
              "name": "Surname generation",
              "node": {
                "@id": "surname-generation",
                "description": "Surname generation step",
                "name": "Surname generation",
                "output": {
                  "@id": "surnames",
                  "name": "Surnames",
                  "schema": "{
                "type": "object",
                "properties": {
                  "surnames": {
                    "type": "array",
                    "items": {
                      "type": "string"
                    }
                  }
                }
              }",
                },
                "prompt": "Produce \${surnamesLength} surnames that sound good with the name: \${name}",
              },
              "output": Output {
                "description": undefined,
                "name": "Surnames",
                "output": {
                  "@id": "surnames",
                  "name": "Surnames",
                  "schema": "{
                "type": "object",
                "properties": {
                  "surnames": {
                    "type": "array",
                    "items": {
                      "type": "string"
                    }
                  }
                }
              }",
                },
                "schema": {
                  "properties": {
                    "surnames": {
                      "items": {
                        "type": "string",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
              },
              "parent": [Circular],
              "prompt": Prompt {
                "prompt": "Produce \${surnamesLength} surnames that sound good with the name: \${name}",
                "variables": [
                  "\${surnamesLength}",
                  "\${name}",
                ],
              },
              "step": {
                "@id": "surname-generation",
                "description": "Surname generation step",
                "name": "Surname generation",
                "output": {
                  "@id": "surnames",
                  "name": "Surnames",
                  "schema": "{
                "type": "object",
                "properties": {
                  "surnames": {
                    "type": "array",
                    "items": {
                      "type": "string"
                    }
                  }
                }
              }",
                },
                "prompt": "Produce \${surnamesLength} surnames that sound good with the name: \${name}",
              },
            },
            Step {
              "description": "Surname selection step",
              "input": undefined,
              "name": "Surname selection",
              "node": {
                "@id": "surname-selection",
                "description": "Surname selection step",
                "name": "Surname selection",
                "output": {
                  "@id": "surname",
                  "name": "Surname",
                  "schema": "{
                "type": "object",
                "properties": {
                  "surname": {
                    "type": "string"
                  }
                }
              }",
                },
                "prompt": "Given the name: \${name} select one of the surnames: \${surnames}",
              },
              "output": Output {
                "description": undefined,
                "name": "Surname",
                "output": {
                  "@id": "surname",
                  "name": "Surname",
                  "schema": "{
                "type": "object",
                "properties": {
                  "surname": {
                    "type": "string"
                  }
                }
              }",
                },
                "schema": {
                  "properties": {
                    "surname": {
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
              "parent": [Circular],
              "prompt": Prompt {
                "prompt": "Given the name: \${name} select one of the surnames: \${surnames}",
                "variables": [
                  "\${name}",
                  "\${surnames}",
                ],
              },
              "step": {
                "@id": "surname-selection",
                "description": "Surname selection step",
                "name": "Surname selection",
                "output": {
                  "@id": "surname",
                  "name": "Surname",
                  "schema": "{
                "type": "object",
                "properties": {
                  "surname": {
                    "type": "string"
                  }
                }
              }",
                },
                "prompt": "Given the name: \${name} select one of the surnames: \${surnames}",
              },
            },
            Step {
              "description": "Character generation step",
              "input": undefined,
              "name": "Character generation",
              "node": {
                "@id": "character-generation",
                "description": "Character generation step",
                "name": "Character generation",
                "output": {
                  "@id": "character",
                  "name": "Character",
                  "schema": "{
                "type": "object",
                "properties": {
                  "character": {
                    "type": "string"
                  }
                }
              }",
                },
                "prompt": "Generate a character with the name: \${name} and the surname: \${surname}",
              },
              "output": Output {
                "description": undefined,
                "name": "Character",
                "output": {
                  "@id": "character",
                  "name": "Character",
                  "schema": "{
                "type": "object",
                "properties": {
                  "character": {
                    "type": "string"
                  }
                }
              }",
                },
                "schema": {
                  "properties": {
                    "character": {
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
              },
              "parent": [Circular],
              "prompt": Prompt {
                "prompt": "Generate a character with the name: \${name} and the surname: \${surname}",
                "variables": [
                  "\${name}",
                  "\${surname}",
                ],
              },
              "step": {
                "@id": "character-generation",
                "description": "Character generation step",
                "name": "Character generation",
                "output": {
                  "@id": "character",
                  "name": "Character",
                  "schema": "{
                "type": "object",
                "properties": {
                  "character": {
                    "type": "string"
                  }
                }
              }",
                },
                "prompt": "Generate a character with the name: \${name} and the surname: \${surname}",
              },
            },
          ],
          "workflow": {
            "@id": "character-building",
            "description": "A workflow that allows you to build a character with a given name",
            "input": {
              "description": "Input description",
              "name": "Human name",
              "schema": "{
            "type": "object",
            "properties": {
              "name": {
                "type": "string"
              },
              "surnamesLength": {
                "type": "number"
              }
            }
          }",
            },
            "name": "Character building",
            "steps": {
              "step": [
                {
                  "@id": "surname-generation",
                  "description": "Surname generation step",
                  "name": "Surname generation",
                  "output": {
                    "@id": "surnames",
                    "name": "Surnames",
                    "schema": "{
                "type": "object",
                "properties": {
                  "surnames": {
                    "type": "array",
                    "items": {
                      "type": "string"
                    }
                  }
                }
              }",
                  },
                  "prompt": "Produce \${surnamesLength} surnames that sound good with the name: \${name}",
                },
                {
                  "@id": "surname-selection",
                  "description": "Surname selection step",
                  "name": "Surname selection",
                  "output": {
                    "@id": "surname",
                    "name": "Surname",
                    "schema": "{
                "type": "object",
                "properties": {
                  "surname": {
                    "type": "string"
                  }
                }
              }",
                  },
                  "prompt": "Given the name: \${name} select one of the surnames: \${surnames}",
                },
                {
                  "@id": "character-generation",
                  "description": "Character generation step",
                  "name": "Character generation",
                  "output": {
                    "@id": "character",
                    "name": "Character",
                    "schema": "{
                "type": "object",
                "properties": {
                  "character": {
                    "type": "string"
                  }
                }
              }",
                  },
                  "prompt": "Generate a character with the name: \${name} and the surname: \${surname}",
                },
              ],
            },
          },
        },
      ],
    },
    Skill {
      "description": "Building a movie scenario means
            creating a story that takes place in
            a world with certain characters, 
            plot lines and themes. 
            The world can be a fantasy world, 
            a historical period, a real world city, 
            or any other setting imaginable. 
            The story can be a romance, a thriller, 
            a sci-fi epic or any other genre. The characters can be humans, animals, aliens, or any other type of being. The plot can be a simple one, like a romance that takes place in a small town, or a complex one, like a multi-layered epic that takes place in a fictional world. The themes can be simple, like love and friendship, or complex, like the nature of reality and the human condition. The movie scenario building skill allows you to explore all of these possibilities and create a unique story that expresses your imagination and creativity.",
      "name": "Movie building skill",
      "node": {
        "description": "Building a movie scenario means
            creating a story that takes place in
            a world with certain characters, 
            plot lines and themes. 
            The world can be a fantasy world, 
            a historical period, a real world city, 
            or any other setting imaginable. 
            The story can be a romance, a thriller, 
            a sci-fi epic or any other genre. The characters can be humans, animals, aliens, or any other type of being. The plot can be a simple one, like a romance that takes place in a small town, or a complex one, like a multi-layered epic that takes place in a fictional world. The themes can be simple, like love and friendship, or complex, like the nature of reality and the human condition. The movie scenario building skill allows you to explore all of these possibilities and create a unique story that expresses your imagination and creativity.",
        "name": "Movie building skill",
      },
      "parent": [Circular],
      "skill": {
        "description": "Building a movie scenario means
            creating a story that takes place in
            a world with certain characters, 
            plot lines and themes. 
            The world can be a fantasy world, 
            a historical period, a real world city, 
            or any other setting imaginable. 
            The story can be a romance, a thriller, 
            a sci-fi epic or any other genre. The characters can be humans, animals, aliens, or any other type of being. The plot can be a simple one, like a romance that takes place in a small town, or a complex one, like a multi-layered epic that takes place in a fictional world. The themes can be simple, like love and friendship, or complex, like the nature of reality and the human condition. The movie scenario building skill allows you to explore all of these possibilities and create a unique story that expresses your imagination and creativity.",
        "name": "Movie building skill",
      },
      "workflows": undefined,
    },
  ],
  "workflows": [
    Workflow {
      "description": "Workflow 2 description",
      "name": "Movie ideas",
      "node": {
        "@id": "movie-ideas",
        "description": "Workflow 2 description",
        "input": {
          "description": "The genre of the movie",
          "name": "Movie idea genre",
          "schema": "{
                        "type": "object",
                        "properties": {
                            "genre": {
                                "type": "string"
                            }
                        }
                    }",
        },
        "name": "Movie ideas",
        "steps": {
          "step": [
            {
              "@id": "ideas-generation",
              "output": {
                "@id": "ideas",
                "name": "Movie ideas",
                "schema": "{
                                "type": "object",
                                "properties": {
                                    "ideas": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    }
                                }
                            }",
              },
              "prompt": "generate 10 movie ideas for \${genre}. 
                    Create a story that takes place in a 
                    world with certain characters, plot lines and themes.",
            },
          ],
        },
      },
      "parent": undefined,
      "steps": [
        Step {
          "description": undefined,
          "input": undefined,
          "name": undefined,
          "node": {
            "@id": "ideas-generation",
            "output": {
              "@id": "ideas",
              "name": "Movie ideas",
              "schema": "{
                                "type": "object",
                                "properties": {
                                    "ideas": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    }
                                }
                            }",
            },
            "prompt": "generate 10 movie ideas for \${genre}. 
                    Create a story that takes place in a 
                    world with certain characters, plot lines and themes.",
          },
          "output": Output {
            "description": undefined,
            "name": "Movie ideas",
            "output": {
              "@id": "ideas",
              "name": "Movie ideas",
              "schema": "{
                                "type": "object",
                                "properties": {
                                    "ideas": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    }
                                }
                            }",
            },
            "schema": {
              "properties": {
                "ideas": {
                  "items": {
                    "type": "string",
                  },
                  "type": "array",
                },
              },
              "type": "object",
            },
          },
          "parent": undefined,
          "prompt": Prompt {
            "prompt": "generate 10 movie ideas for \${genre}. 
                    Create a story that takes place in a 
                    world with certain characters, plot lines and themes.",
            "variables": [
              "\${genre}",
            ],
          },
          "step": {
            "@id": "ideas-generation",
            "output": {
              "@id": "ideas",
              "name": "Movie ideas",
              "schema": "{
                                "type": "object",
                                "properties": {
                                    "ideas": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    }
                                }
                            }",
            },
            "prompt": "generate 10 movie ideas for \${genre}. 
                    Create a story that takes place in a 
                    world with certain characters, plot lines and themes.",
          },
        },
      ],
      "workflow": {
        "@id": "movie-ideas",
        "description": "Workflow 2 description",
        "input": {
          "description": "The genre of the movie",
          "name": "Movie idea genre",
          "schema": "{
                        "type": "object",
                        "properties": {
                            "genre": {
                                "type": "string"
                            }
                        }
                    }",
        },
        "name": "Movie ideas",
        "steps": {
          "step": [
            {
              "@id": "ideas-generation",
              "output": {
                "@id": "ideas",
                "name": "Movie ideas",
                "schema": "{
                                "type": "object",
                                "properties": {
                                    "ideas": {
                                        "type": "array",
                                        "items": {
                                            "type": "string"
                                        }
                                    }
                                }
                            }",
              },
              "prompt": "generate 10 movie ideas for \${genre}. 
                    Create a story that takes place in a 
                    world with certain characters, plot lines and themes.",
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`assistant starts workflow 1`] = `
[
  {
    "data": {
      "surnames": [
        "Doe",
        "Smith",
      ],
    },
    "prompt": "Produce 2 surnames that sound good with the name: John",
  },
  {
    "data": {
      "surname": "Doe",
    },
    "prompt": "Given the name: John select one of the surnames: Doe,Smith",
  },
  {
    "data": {
      "character": "John Doe",
    },
    "prompt": "Generate a character with the name: John and the surname: Doe",
  },
]
`;
